database/tbcd/level/level.go

```diff
@@ -1707,6 +1729,66 @@ func (l *ldb) BlockTxUpdate(ctx context.Context, direction int, txs map[tbcd.TxK
 	return nil
 }
 
+func (l *ldb) BlockKeystoneUpdate(ctx context.Context, direction int, keystones map[chainhash.Hash]tbcd.Keystone) error {
+	log.Tracef("BlockKeystoneUpdate")
+	defer log.Tracef("BlockKeystoneUpdate exit")
+
+	if !(direction == 1 || direction == -1) {
+		return fmt.Errorf("invalid direction: %v", direction)
+	}
+
+	// keystones
+	kssTx, kssCommit, kssDiscard, err := l.startTransaction(level.KeystonesDB)
+	if err != nil {
+		return fmt.Errorf("keystones open db transaction: %w", err)
+	}
+	defer kssDiscard()
+
+	kssBatch := new(leveldb.Batch)
+	for k, v := range keystones {
+		switch direction {
+		case -1:
+			foundKeystone, err := kssTx.Get(k[:], nil)
+			if err != nil {
+				continue
+			}
+			fbh, err := chainhash.NewHash(foundKeystone[0:32])
+			if err != nil {
+				return err // XXX really can't happen
+			}
+			// Only delete keystone if it is in the previously found block.
+			if fbh.IsEqual(&v.BlockHash) {
+				kssBatch.Delete(k[:])
+			}
+		case 1:
+			has, err := kssTx.Has(k[:], nil)
+			if err != nil {
+				return fmt.Errorf("keystone update has: %w", err)
+			}
+			if has {
+				// Only store unknown keystones
+				continue
+			}
+			var value [chainhash.HashSize + hemi.L2KeystoneAbrevSize]byte
```

Comment:
let's assert that `(len(v.BlockHash) + len(v.AbbreviatedKeystone) == chainhash.HashSize + hemi.L2KeystoneAbrevSize`

if not, we could get data corruption here, which would be bad

Reply:
These are all chainhash.Hash that are forced to 32 bytes.